gms/gossiper.cc

```diff
@@ -2150,27 +2275,38 @@ future<> gossiper::wait_for_gossip(std::chrono::milliseconds initial_delay, std:
 }
 
 future<> gossiper::wait_for_gossip_to_settle() {
-    static constexpr std::chrono::milliseconds GOSSIP_SETTLE_MIN_WAIT_MS{5000};
-    auto force_after = _cfg.skip_wait_for_gossip_to_settle();
+    logger.info("Waiting for gossip to settle with query...");
     auto do_enable_features = [this] {
         return async([this] {
             if (!std::exchange(_gossip_settled, true)) {
                maybe_enable_features();
             }
         });
     };
+    auto force_after = _cfg.skip_wait_for_gossip_to_settle();
     if (force_after == 0) {
         return do_enable_features();
     }
-    return wait_for_gossip(GOSSIP_SETTLE_MIN_WAIT_MS, force_after).then([this, do_enable_features] {
+    return wait_for_cluster_ring_status_to_match(*this).then([do_enable_features] {
         return do_enable_features();
+    }).handle_exception([this, do_enable_features] (std::exception_ptr ep) {
+        logger.debug("Failed to run wait_for_cluster_ring_status_to_match: {}", ep);
+        static constexpr std::chrono::milliseconds GOSSIP_SETTLE_MIN_WAIT_MS{5000};
+        logger.info("Waiting for gossip to settle without query...");
+        return wait_for_gossip(GOSSIP_SETTLE_MIN_WAIT_MS).then([this, do_enable_features] {
+            return do_enable_features();
+        });
     });
 }
 
 future<> gossiper::wait_for_range_setup() {
-    logger.info("Waiting for pending range setup...");
-    auto ring_delay = std::chrono::milliseconds(_cfg.ring_delay_ms());
-    return wait_for_gossip(ring_delay);
+    logger.info("Waiting for pending range setup with query...");
+    return wait_for_cluster_ring_status_to_match(*this).handle_exception([this] (std::exception_ptr ep) {
+        logger.debug("Failed to run wait_for_cluster_ring_status_to_match: {}", ep);
```

Comment:
No. It is legal to fall back to the time based methods.

Reply:
But is it legal to fall back? We call `wait_for_range_setup()` when we join a new node to the cluster. So we can't have nodes behind a network partition not learning about our tokens, or they will send writes to wrong replicas. Otherwise this PR is not really making anything safer, just faster on the happy path.

Reply:
@kbr- Please. This patch does not solve how to make adding node procedure safe. The title does not say it makes *bootstrap* safer. It says it makes boot process safer because with this patch, we use hash and not timeout to make sure peers see the tokens. Let's not discuss adding node procedure here.

Reply:
This patch changes the bootstrap as well though. In the happy path, it reduces the ring delay wait which makes https://github.com/scylladb/scylla/issues/5251 and https://github.com/scylladb/scylla/issues/5252 more likely. Therefore I think we should address those issues before we get rid of ring delay wait during topology change operations.

Reply:
@asias 

> It says it makes boot process safer

So you mean it just makes restart safer?

> we use hash and not timeout to make sure peers see the tokens.

If I understand the patch, we don't use hash to **make sure**. We only **try** to use the hash and if we fail, we fall-back to timeout based approach.

Reply:
It also make bootstrap safer, e.g., use the hash to make sure the peer node know the bootstrap node's token. 

Yes, you can say that when it falls back to timeout method it is not safe. It fallbacks when node is down.